### PR TITLE
sds_validation: fix sub params and test case issues

### PR DESCRIPTION
**Pull request by TimothyAsir**
1) As per the latest change in the schema, Some object can be defined directly without specifying its parent name. Means, any mandatory or optional parameter can be specified with out parent name for the whole object. i.e:- parent_object_name = parameter_name

      inputs:
        mandatory:
          - "Node[]"
          - Tendrl_context.sds_name
          - Tendrl_context.sds_version

Currently sds-validation does not support any parameter to be specified without the object type.
For example if the parameter is specified as Volume.brickdetails or Node.Node it will validate; where as, sds-validation skips the parameter check if the particular parameter is specified without the parent object like Node or Node[] which may have a single element or a list of elements

This patch (line 234-241) fixes this issue by making parent_object_name = parameter_name. Because there is no parent for this.

2) Input parameter is case sensitive
Currently sds-validation will convert the custom type object name to lower case (line no 279). Which may cause "custom object not defined" error if the object is defined with upper case.
This patch fixes this case sensitive issue by removing the to_lower_case conversion.

mandatory:
              - Volume.volname
              - Volume.brickdetails

3) Input parameter can have nested input parameter (dict of dict of ...)
Currently sds-validate will not support multi level and support only one sub level. Which will not help in case of input parameters have multiple sub parameters. For example in case of importNode the parameter can have nested input in addition to its regular inputs like Node: [ { fqdn, ...} ].

This patch [line no. 293 - 309] fixes this issue by running into the sub objects and go to the lease element in the chain and checks its parameter type recursively for every items and checks the whole object.

4) Every objects have a name-space defined
Currently sds-validation test cases could not succeeded due to the sds-schema changes to introduce name-space details.

5) Removed __init__.py file in tests directory which is not used and causes some test case failures
during rpmbuild. Suggested by http://doc.pytest.org/en/latest/goodpractices.html

This patch fixes the above 5 issues.

Api Job can be any as follows
~~~~~~~~~~~~~~~~~~~~~~~~

 "parameters": {
                'sds_version': "3.0",
                'sds_name': "node",
                'Node': [{'fqdn': "abc.12.com", 'cmd_str': 'install'},
                         {'fqdn': "xyz.com", 'cmd_str': 'execute'}]
            }

or
 "parameters": {
                'sds_version': "3.0",
                'sds_name': "node",
                'Node[]': [{'fqdn': "abc.12.com", 'cmd_str': 'install'},
                         {'fqdn': "xyz.com", 'cmd_str': 'execute'}]
            }

or

 "parameters": {
                'sds_version': "3.0",
                'sds_name': "node",
                'Node': {'fqdn': "abc.12.com", 'cmd_str': 'install'},
            }


Signed-off-by: Timothy Asir J <tjeyasin@redhat.com>

### CODE REVIEW

**Comment by TimothyAsir** [comment]
1) As per the latest change in the schema, Some object can be defined directly without specifying its parent name. Means, any mandatory or optional parameter can be specified with out parent name for the whole object. i.e:- parent_object_name = parameter_name

Currently sds-validation does not support any parameter to be specified without the object type.
For example if the parameter is specified as Volume.brickdetails or Node.Node it will validate;  where as, sds-validation skips the parameter check if the particular parameter is specified without the parent object like Node or Node[] which may have a single element or a list of elements

This patch (line 234-241) fixes this issue by making parent_object_name = parameter_name. Because there is no parent for this.

2) Input parameter is case sensitive
Currently sds-validation will convert the custom type object name to lower case (line no 279). Which may cause "custom object not defined" error if the object is defined with upper case.
This patch fixes this case sensitive issue by removing the to_lower_case conversion.

3) Input parameter can have nested input parameter (dict of dict of ...)
Currently sds-validate will not support multi level and support only one sub level. Which will not help in case of input parameters have multiple sub parameters. For example in case of importNode the parameter can have nested input in addition to its regular inputs like Node: [ { fqdn, ...} ].

This patch [line no. 293 - 309] fixes this issue by running into the sub objects and go to the lease element in the chain and checks its parameter type recursively for every items and checks the whole object.

4) Every objects have a name-space defined
Currently sds-validation test cases could not succeeded due to the sds-schema changes to introduce name-space details.
This patch fixes this issue.

Other known minor issues to be fixed which will not cause any issues currently.
1) sds-schema namespace can have flows without any objects
2) how are we going to deal with sub object mandatory and optional parameters.
for example, Let Node object has attrs and atoms. How to specify the optional and mandatory parameters if Node itself going to be a parameter or an api job.

**Comment by TimothyAsir** [comment]
@r0h4n , @shtripat , @nthomas-redhat , Please review

File: tendrl/common/tests/sds_state_gluster.yaml [hunk 1]
```diff
@@ -1,13 +1,7 @@
...
-tendrl_schema_version: 0.1
-valid_objects:
-- tendrl_context
-- volume
-- brick
-- peer
-
-object_details:
-  tendrl_context:
+namespace.tendrl.gluster_bridge:
```
Comment (shtripat) [inline tendrl/common/tests/sds_state_gluster.yaml]: namespace.tendrl.gluster_integration

File: tendrl/common/tests/sds_state_gluster.yaml [hunk 2]
```diff
@@ -47,7 +41,7 @@ object_details:
         type: String
         help: "Transport type for volume"
       brickdetails:
-        type: brick[]
```
Comment (shtripat) [inline tendrl/common/tests/sds_state_gluster.yaml]: brickdetails is renamed as bricks now

File: tendrl/common/tests/sds_state_gluster.yaml [hunk 3]
```diff
@@ -58,7 +52,7 @@ object_details:
           enabled: True
           uuid: "242f6190-9b37-11e6-950d-a24fc0d9649c"
           flows: ["CreateVolume"]
-          run: "tendrl.gluster_bridge.atoms.volume.create.Create"
+          run: "tendrl.gluster_bridge.flows.Create"
```
Comment (shtripat) [inline tendrl/common/tests/sds_state_gluster.yaml]: Why flows under atom?

File: tendrl/common/tests/sds_state_gluster.yaml [hunk 4]
```diff
@@ -103,14 +97,15 @@ object_details:
       volname:
         type: String
         help: "Gluster volume name"
-flows:
+namespace.tendrl.gluster_bridge.gluster_integration:
```
Comment (shtripat) [inline tendrl/common/tests/sds_state_gluster.yaml]: Its "namespace.tendrl.gluster_integration" only

File: tendrl/common/tests/sds_state_gluster.yaml [hunk 5]
```diff
@@ -103,14 +97,15 @@ object_details:
...
+ flows:
   CreateGlusterVolume:
     enabled: True
     tendrl_context: "$cluster.id"
     uuid: "1951e821-7aa9-4a91-8183-e73bc8275b8e"
     description: "Create Volume with pre-existing bricks"
     atoms:
-      - volume.atoms.create
+      - tendrl.gluster_bridge.objects.Volume.atoms.create
```
Comment (shtripat) [inline tendrl/common/tests/sds_state_gluster.yaml]: gluster_inetgration??

File: tendrl/common/tests/sds_state_node.yaml [hunk 6]
```diff
@@ -1,74 +1,155 @@
...
-              - general.conf_file
-              - general.status
-        version: "0.0.1"
+        cmd:
+          enabled: true
+          inputs:
+            mandatory:
+              - Node.cmd_str
+          name: "Execute CMD on Node"
+          run: tendrl.node_agent.atoms.node.cmd.Cmd
```
Comment (shtripat) [inline tendrl/common/tests/sds_state_node.yaml]: atoms should not have Class name mentioned

File: tendrl/common/tests/sds_state_node.yaml [hunk 7]
```diff
@@ -1,74 +1,155 @@
...
+        sds_version:
+          help: "Version of the Tendrl managed sds, eg: '3.2.1'"
+          type: String
+      enabled: true
+namespace.tendrl.node_agent.gluster_integration:
+  flows:
+    ImportCluster:
+      atoms:
+        - tendrl.node_agent.objects.Package.atoms.install
+        - tendrl.node_agent.gluster_integration.objects.Config.atoms.generate
```
Comment (shtripat) [inline tendrl/common/tests/sds_state_node.yaml]: package names are snakecase and not camelCase

File: tendrl/common/tests/test_validate_api.py [hunk 8]
```diff
@@ -8,16 +8,17 @@
 import os
 import sys
 
-sys.path.insert(0, '../../')
-from tendrl.common.definitions.validator import JobValidator
-from tendrl.common import util
+#sys.path.insert(0, '../')
+sys.path.append('../..')
+from bridge_common.definitions.validator import JobValidator
```
Comment (shtripat) [inline tendrl/common/tests/test_validate_api.py]: bridge_common is now renamed as "common"

File: tendrl/common/tests/sds_state_gluster.yaml [hunk 9]
```diff
@@ -1,13 +1,7 @@
 ---
```
Comment (r0h4n) [inline tendrl/common/tests/sds_state_gluster.yaml]: why the name "sds_state_gluster.yaml"? use "tendrl_definitions_gluster.yaml"

File: tendrl/common/tests/sds_state_node.yaml [hunk 10]
```diff
@@ -1,74 +1,155 @@
 ---
```
Comment (r0h4n) [inline tendrl/common/tests/sds_state_node.yaml]: why the name "sds_state_node.yaml"? use "tendrl_definitions_node.yaml"

File: tendrl/common/tests/sds_state_gluster.yaml [hunk 11]
```diff
@@ -58,7 +52,7 @@ object_details:
           enabled: True
           uuid: "242f6190-9b37-11e6-950d-a24fc0d9649c"
           flows: ["CreateVolume"]
-          run: "tendrl.gluster_bridge.atoms.volume.create.Create"
+          run: "tendrl.gluster_bridge.flows.Create"
           type: "create"
           pre_run:
             - "tendrl.gluster_bridge.atoms.volume.validations.volume_not_exists"
```
Comment (r0h4n) [inline tendrl/common/tests/sds_state_gluster.yaml]: Update this yaml from the latest yaml, we no longer use "gluster_bridge"

File: tendrl/common/tests/sds_state_node.yaml (hunk 10 above)
Comment (r0h4n) [inline tendrl/common/tests/sds_state_node.yaml]: Update this yaml, it is old

File: tendrl/common/tests/test_validate_api.py [hunk 12]
```diff
@@ -148,8 +152,8 @@ def test_apijob_with_wrong_argument_name(self):
             'cluster_id': "49fa2adde8a6e98591f0f5cb4bc5f44d",
             "sds_name": "gluster",
             "sds_version": "3.2.0",
-            "flow": "CreateGlusterVolume",
-            "status": "processing",
+            "flow": "namespace.tendrl.gluster_bridge.gluster_integration.flows.CreateGlusterVolume",
```
Comment (r0h4n) [inline tendrl/common/tests/test_validate_api.py]: "flow" is removed, it is now "run" , and without the "namespace" prefix

File: tendrl/common/tests/test_validate_api.py [hunk 13]
```diff
@@ -167,20 +171,17 @@ def test_apijob_with_wrong_argument_name(self):
 
     def test_apijob_missing_argument(self):
         nodeApiJob = {
-            'cluster_id': "49fa2adde8a6e98591f0f5cb4bc5f44d",
-            "node_name": "fqdn",
-            "node_version": "3.2.0",
-            "flow": "PackageInstall",
+            "cluster_id": "bc35fde8-528b-4908-8d52-50d8bf964dd4",
+            "node_uuid": "node_uuid",
```
Comment (r0h4n) [inline tendrl/common/tests/test_validate_api.py]: "node_uuid" is "node_id"

**Comment by r0h4n** [comment]
Below proposed specs have an impact on common/validation:
https://github.com/Tendrl/specifications/pulls/r0h4n

Please ensure the impact is well understood and then update this PR

**Pull request closed without merging**